hadoop-plugin/src/main/groovy/com/linkedin/gradle/zip/HadoopZipExtension.groovy

```diff
@@ -204,6 +209,55 @@ class HadoopZipExtension {
     }
   }
 
+  /**
+   * This enables "Emergent Flows" in Flow 2.0 - so users can define properties in a separate
+   * namespace then merge them into the resulting yaml workflows at Zip time.
+   *
+   * This method retrieves all .tempprops source files in the task and turns them into Maps, then
+   * merges those maps into the root level config of all of the .flow file (to ensure backward
+   * compatibility). The resulting object is then saved in a new .flow file.
+   *
+   * @param task Task where merging may occur
+   * @param zipName Name of the zip being generated - used to create new .flow file names
+   */
+  void mergeTempPropsWithFlowFiles(Task task, String zipName) {
+    List<Map<String, String>> tempPropsList = [];
+    List<String> tempPropsPathsList = [];
+    List<String> flowFilePathsList = [];
+    List<String> newFlowFilePathsList = [];
+
+    task.source.visit { FileVisitDetails tempProps ->
+      String tempPropsPath = tempProps.path;
+      if (tempPropsPath.matches(/.*\.tempprops/)) {
+        Map<String, String> tempPropsMap = YamlMerge.readInYaml(tempProps);
+        tempPropsList.add(tempPropsMap);
+        tempPropsPathsList.add(tempPropsPath);
+      }
+    }
+    // Merge all .tempprops files into the .flow files and create new .flow files with the results
+    if (!tempPropsList.isEmpty()) {
+      task.source.visit { FileVisitDetails file ->
+        String flowFile = file.path;
+        if (flowFile.matches(/.*\.flow/) && !flowFile.matches(/.*_.*/)) {
```

Comment:
Can the name of a flow file before merging also contain underscore? e.g. "jasun_test.flow"

Reply:
Good question - no it cannot. Flow names (and all object names in the DSL in general) can only contain letters, numbers, and dashes.

Reply:
I remember the job file names in Flow 1.0 allow using "_". Is it some new requirement for the flow names in Flow 2.0?  Can you show me where the code is? Thanks!

Reply:
Discussed briefly offline. Job file names in Flow 1.0 only contained "_" after being processed by the DSL, users could not declare a job with name "my_test".

This did remind me in the else if block to also check if the file is a .flow file, which apparently I had forgotten. Adding in the next commit - thanks!